Source/Core/Common/BitFieldView.h

```diff
@@ -0,0 +1,788 @@
+  AssertBitFieldFixedIsSane<field_t, start, width, host_t>();
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  using shost_t = std::make_signed_t<host_t>;
+  constexpr std::size_t rshift = 8 * sizeof(host_t) - width;
+  constexpr std::size_t lshift = rshift - start;
+
+  // Choose between arithmetic (sign extend) or logical (no sign extend) right-shift.
+  // Fun fact: This is implementaton defined.  It is only possible to do this implicitly.
+  if constexpr (std::is_signed_v<field_t>)
+    return static_cast<field_t>(static_cast<shost_t>(host << lshift) >> rshift);
+  else
+    return static_cast<field_t>(static_cast<uhost_t>(host << lshift) >> rshift);
+}
+
+template <typename field_t, typename host_t>
+constexpr field_t GetBitField(const std::size_t start, const std::size_t width, const host_t host)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldIsSane<field_t, host_t>(start, width);
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  using shost_t = std::make_signed_t<host_t>;
+  const std::size_t rshift = 8 * sizeof(host_t) - width;
+  const std::size_t lshift = rshift - start;
+
+  // Choose between arithmetic (sign extend) or logical (no sign extend) right-shift.
+  // Fun fact: This is implementaton defined.  It is only possible to do this implicitly.
+  if constexpr (std::is_signed_v<field_t>)
+    return static_cast<field_t>(static_cast<shost_t>(host << lshift) >> rshift);
+  else
+    return static_cast<field_t>(static_cast<uhost_t>(host << lshift) >> rshift);
+}
+
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+constexpr void SetBitFieldFixed(host_t& host, const field_t val)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldFixedIsSane<field_t, start, width, host_t>();
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  constexpr std::size_t rshift = 8 * sizeof(host_t) - width;
+  constexpr uhost_t mask = std::numeric_limits<uhost_t>::max() >> rshift << start;
+
+  host = (host & ~mask) | ((static_cast<uhost_t>(val) << start) & mask);
+}
+
+template <typename field_t, typename host_t>
+constexpr void SetBitField(const std::size_t start, const std::size_t width, host_t& host,
+                           const field_t val)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldIsSane<field_t, host_t>(start, width);
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  const std::size_t rshift = 8 * sizeof(host_t) - width;
+  const uhost_t mask = std::numeric_limits<uhost_t>::max() >> rshift << start;
+
+  host = (host & ~mask) | ((static_cast<uhost_t>(val) << start) & mask);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Float host type specialization
+
+template <typename field_t, std::size_t start, std::size_t width>
+constexpr field_t GetBitFieldFixed(const float& host)
```

Comment:
Why is the argument type `const float&`? Ditto below, and for `const double&` instances.

Reply:
This can likely be simplified to:
```cpp
template <typename field_t, std::size_t start, std::size_t width>
constexpr field_t GetBitFieldFixed(float host)
{
    return GetBitFieldFixed<field_t, start, width>(Common::BitCast<std::uint32_t>(host));
}
```
This would be slightly more DRY, as well as being easier to read.

Reply:
The floating point specializations use BitCast, which accepts an lvalue reference type, in order to perform bitwise operations on the host value.  The original template does not need a reference (a pointer) to the host value in memory in order to perform bitwise operations on it.  The idea is that the value remains in a register for the templated version, and is more performant that way.  In short, any specializations where the host value must exist in memory shall accept the host by lvalue reference rather than by value.
I do like your DRY suggestion, thanks.

Reply:
I just now almost considered adding the register specifier to the templated GetBitField/Fixed function's host_t parameter, only to learn it has been removed as of C++17!  It would have been nice to self-document the intent of passing by value rather than lvalue reference, but emitting a compiler warning isn't worth that.